### PR TITLE
Refactor CUDA qjit decorator to inherit from catalyst.QJIT

### PR DESCRIPTION
**Pull request by josh146**
**Context:** With the change in #531, we can now modify the CUDA `@qjit` decorator to inherit from the main `catalyst.qjit` decorator. This enables the functionality provided by the main QJIT class to be supported by the CUDA quantum QJIT, while reducing code duplication.

**Description of the Change:**

- Modifies `QJIT_CUDA` to inherit from `catalyst.QJIT`
- `catalyst.cuda.cudaqjit` is rewritten to match the `catalyst.qjit` wrapper function implementation.
- the CUDA `interpret` function is now redundant and can be removed.

**Benefits:**

- Support for caching the JAXPR capture --- this happens 'just in time' (or ahead of time), and JAXPR capture is not repeated on every execution of the kernel.
- autograph support is now enabled

**Possible Drawbacks:**

- autograph support won't work end-to-end until CUDA Q support for for loops and if statements is added.
- `static_argnums` support is _not_ enabled, I'm not sure if this makes sense here, given the CUDA qjit is simply doing just-in-time _capture_, and not just-in-time compilation.
- Similarly, other qjit keyword arguments such as pipeline, target, verbose, etc. likely also do not make sense/cannot be supported, and are hardcoded to be disabled.
- The `extract_backend_info` should not be automatically run by the cuda compiler as it is
catalyst-specific. We need to make this API a bit nicer for third-party compilers.

**Related GitHub Issues:** n/a

### CODE REVIEW

**Comment by josh146** [comment]
The only test not currently working is `test_expval_2`:

```python
@cudaq_qjit
@qml.qnode(qml.device("softwareq.qpp", wires=2))
def circuit():
    qml.RY(jnp.pi / 4, wires=[1])
    return qml.expval(qml.PauliZ(1) + qml.PauliX(1))
```

<details>
<summary>See here for the traceback</summary>

```pycon
>>> circuit()
[/usr/local/lib/python3.10/dist-packages/catalyst/jit.py](https://localhost:8080/#) in __call__(self, *args, **kwargs)
    118             args = promote_arguments(self.c_sig, dynamic_args)
    119 
--> 120         return self.run(args, kwargs)
    121 
    122     def aot_compile(self):

[<ipython-input-15-41f6ffd49162>](https://localhost:8080/#) in run(self, args, kwargs)
     52         # # use catalyst.cuda.cuda_qjit to interpret the JAXPR and execute via CUDA Quantum
     53         ctx = InterpreterContext(self.jaxpr.jaxpr, self.jaxpr.literals, *args)
---> 54         results = interpret_impl(ctx, self.jaxpr.jaxpr)
     55 
     56         return tree_unflatten(self.out_treedef, results)

[/usr/local/lib/python3.10/dist-packages/catalyst/cuda/catalyst_to_cuda_interpreter.py](https://localhost:8080/#) in interpret_impl(ctx, jaxpr)
    785         # This is similar to direct-call threading
    786         # https://www.cs.toronto.edu/~matz/dissertation/matzDissertation-latex2html/node6.html
--> 787         INST_IMPL.get(eqn.primitive, default_impl)(ctx, eqn)
    788 
    789     retvals = _map(ctx.read, jaxpr.outvars)

[/usr/local/lib/python3.10/dist-packages/catalyst/cuda/catalyst_to_cuda_interpreter.py](https://localhost:8080/#) in change_hamiltonian(ctx, eqn)
    631     assert eqn.primitive == hamiltonian_p
    632 
--> 633     invals = _map(ctx.read, eqn.invars)
    634     coeffs = invals[0]
    635     terms = invals[1:]

[/usr/local/lib/python3.10/dist-packages/catalyst/cuda/catalyst_to_cuda_interpreter.py](https://localhost:8080/#) in _map(f, *collections)
    146 def _map(f, *collections):
    147     """Eager implementation of map."""
--> 148     return list(map(f, *collections))
    149 
    150 

[/usr/local/lib/python3.10/dist-packages/catalyst/cuda/catalyst_to_cuda_interpreter.py](https://localhost:8080/#) in read(self, var)
    210         if self.variable_map.get(var):
    211             var = self.variable_map[var]
--> 212         return self.env[var]
    213 
    214     def write(self, var, val):

KeyError: a
```
</details>

File: frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py [hunk 1]
```diff
@@ -832,75 +818,38 @@ def cudaq_backend_info(device, _config) -> BackendInfo:
...
-        ctx = InterpreterContext(closed_jaxpr.jaxpr, closed_jaxpr.literals, *args)
-        out = interpret_impl(ctx, closed_jaxpr.jaxpr)
+    def run(self, args, kwargs):
+        # We need to override run, because it currently tries to call self.compiled_function,
+        # which does not exist here.
 
-        out = tree_unflatten(out_tree, out)
-        return out
+        ctx = InterpreterContext(self.jaxpr.jaxpr, self.jaxpr.literals, *args)
+        results = interpret_impl(ctx, self.jaxpr.jaxpr)
```
Comment (erick-xanadu) [inline frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py:853]: Have you tested kwargs? Not sure if the following is all that is required.

```python
flat_args = tree_flatten((args, kwargs))
ctx = InterpreterContext(self.jaxpr.jaxpr, self.jaxpr.literals, *flat_args)
results = interpret_impl(ctx, self.jaxpr.jaxpr)
```

**Comment by erick-xanadu** [comment]
The errors I am seeing on CI are different :thinking: 

```python
 FAILED frontend/test/pytest/test_cuda_integration.py::TestCudaQ::test_qjit_cuda_remove_host_context - catalyst.utils.exceptions.CompileError: Cannot translate tapes with context.
FAILED frontend/test/pytest/test_cuda_integration.py::TestCudaQ::test_expval_2 - KeyError: a
FAILED frontend/test/pytest/test_cuda_integration.py::TestCudaQ::test_jit_capture - AssertionError: Expected 'capture' to not have been called. Called 1 times.
Calls: [call((Array([0.1, 0.2], dtype=float64),))].
FAILED frontend/test/pytest/test_cuda_integration.py::TestCudaQ::test_aot_capture - AssertionError: Expected 'capture' to have been called.
```

File: frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py (hunk 1 above)
Comment (josh146) [reply frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py:853]: I haven't, thank!

Comment (josh146) [reply frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py:853]: Unfortunately this doesn't seem to work :( I've added a kwargs test, but it is marked as skipped currently.

**Comment by josh146** [comment]
> The errors I am seeing on CI are different 🤔

I've just pushed a fix, the only failures should now be `test_expval_2`.

File: frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py [hunk 2]
```diff
@@ -832,75 +818,38 @@ def cudaq_backend_info(device, _config) -> BackendInfo:
...
+        return False  # no type promotion
 
-        # Because they become args...
-        args = list(args) + consts
-        args = tuple(args)
-        ctx = InterpreterContext(closed_jaxpr.jaxpr, closed_jaxpr.literals, *args)
-        out = interpret_impl(ctx, closed_jaxpr.jaxpr)
+    def run(self, args, kwargs):
+        # We need to override run, because it currently tries to call self.compiled_function,
+        # which does not exist here.
```
Comment (dime10) [inline frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py:850]: Customizing run works nicely here and is also short 👍 

If we wanted an alternative with better potential integration into the cache system (related to comment above), we could also just store the desired callable into the `self.compiled_function` attribute during the `compile` stage, something that can be invoked as `self.compiled_function(*args, **kwargs)`.

```py
def compile(self):
  def compiled_fn(*args, **kwargs):
    ctx = InterpreterContext(self.jaxpr.jaxpr, self.jaxpr.literals, *args)
    results = interpret_impl(ctx, self.jaxpr.jaxpr)
  return compiled_fn, None
```
(as an example)

File: frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py (hunk 1 above)
Comment (dime10) [inline frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py:853]: @erick-xanadu is passing the same jaxpr twice needed?

**Review by dime10 — COMMENTED** [review]
Nice work 😍

File: frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py [hunk 3]
```diff
@@ -795,31 +794,18 @@ def interpret_impl(ctx, jaxpr):
...
-
-        Args:
-            *args: either the concrete values to be passed as arguments to ``fn`` or abstract values
-
-        Returns:
-            an MLIR module
-        """
+    def capture(self, args):
+        # Need to override the capture method for two reasons:
+        # - We need to patch the allowed devices.
```
[the remark this answers is not in the document]
Comment (josh146) [reply frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py:807]: This is inherited from @erick-xanadu implementation, so I'll defer to him here; but this is related to the existing docstring below:

> The extract_backend_info should not be run by the cuda compiler as it is
catalyst-specific. We need to make this API a bit nicer for third-party compilers.

Ideally, I was hoping that I wouldn't need to override `capture()`. This was _almost_ possible if I moved the `remove_host_context` to `jit_compile()`, but it wouldn't work (would get error messages of '`softwareq.qpp` is not a valid device').

File: frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py [hunk 4]
```diff
@@ -832,75 +818,38 @@ def cudaq_backend_info(device, _config) -> BackendInfo:
...
-def interpret(fun):
-    """Wrapper function that takes a function that will be interpretted with
-    the semantics in interpret_impl.
+    def jit_compile(self, args):
+        # JIT compile needs to be overridden from the base class, for a couple of reasons:
+        #
+        # - It calls self.generate_ir() and self.compile(), none of which are applicable here
+        #   (for now).
+        # - It tries to create a function cache and a CompiledFunction object, which will fail in
+        #   this case, because we have no compiled binary nor shared library.
```
[the remark this answers is not in the document]
Comment (josh146) [reply frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py:835]: Yep, those were the two issues I ran into. I tried creating a 'mock' `CompiledFunction` object which solved the isinstance check, but then I got errors from failed attempts to open a shared library.

File: frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py:807]: We need to `cudaq_backend_info` function to replace `extract_backend_info`. This doesn't need to be in capture, but I found this location to be a good place to do so.

File: frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py (hunk 1 above)
Comment (erick-xanadu) [reply frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py:853]: Probably can be removed! :+1:

**Comment by dime10** [comment]
@josh146 any chance this was close to landing? Sounded like a decent improvement :D

**Comment by josh146** [comment]
@dime10 this was almost ready to go, there was just a single failing test that was beyond my expertise and I couldn't solve

**Comment by erick-xanadu** [comment]
made obsolete by #926

**Pull request closed without merging**